rust/src/utils.rs

```diff
@@ -0,0 +1,321 @@
+    script_sig.insert(0, 16_u8); // prepend length of script
+    let script_sig = ScriptBuf::from(script_sig);
+    let witness = Witness::from(vec![[0_u8; 32].to_vec()]);
+
+    let input = TxIn {
+        previous_output: OutPoint {
+            //txid must be 32 bytes of zeros for coinbase input and vout must be max value 0xffffffff
+            txid: Txid::from_raw_hash(*OtherHash::from_bytes_ref(&[
+                0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                0, 0, 0, 0,
+            ])),
+            vout: 0xffffffff,
+        },
+        script_sig: script_sig,
+        sequence: Sequence(0xffffffff),
+        witness: witness, // Need 32 byte reserved value all zeros if block contains segwit transactions
+    };
+
+    /*
+    cant use because need the first output of the coinbas transaction to be the wTXID commitment
+    let op_return_data = String::from("Hello Rust Class").into_bytes();
+    // seems excessive but couldnt satisfy pushbytes with any other type
+    let op_return_data: &[u8; 16] = op_return_data.as_slice().try_into().expect("error");
+    let op_return_script = ScriptBuf::new_op_return(op_return_data);
+     */
+
+    let output = TxOut {
+        value: Amount::ZERO,
+        script_pubkey: ScriptBuf::new(),
+    };
+
+    let value = Amount::from_btc(3.125).expect("error creating amount for output");
+    let recieve_script = get_recieve_script(); // need to create to recieve script
+
+    let output2 = TxOut {
+        value: value,
+        script_pubkey: recieve_script,
+    };
+
+    Transaction {
+        version: TransactionVersion::ONE,
+        lock_time: LockTime::ZERO,
+        input: vec![input],
+        output: vec![output, output2],
+    }
+}
+
+// Adds coinbase transaction and transactions to make a block template and updates merkle_root(Returned Block.header is ready to be mined)
+pub fn block_template(mut txdata: Vec<Transaction>) -> Block {
+    let coinbase_tx = create_coinbase_transaction();
+    txdata.insert(0, coinbase_tx);
+
+    let header = block_header();
+    let mut block = Block { header, txdata };
+
+
+    let witness_root = block.witness_root().expect("error calculating root");
+    let witness_reserve_value = [0_u8; 32].to_vec();
+    let witness_commitment = Block::compute_witness_commitment(&witness_root, &witness_reserve_value);
+    let mut witness_commitment = witness_commitment.as_byte_array().to_vec();
+    
+    let mut finished = hex::decode("6a24aa21a9ed").expect("error decoding hex");
+
+    finished.append(&mut witness_commitment);
+
+    block.txdata[0].output[0].script_pubkey =
+        ScriptBuf::from_bytes(finished);
+
+
+    let merkle_root = block
+        .compute_merkle_root()
+        .expect("error calculating merkle root ");
+    block.header.merkle_root = merkle_root;
+    
+
+    block
+}
+
+
+// Increments and updates nonce until valid block is found
+pub fn mine_block(block: &mut Block) -> () {
```

Comment:
Not sure I see how this is working. 

When the nonce is updated, this will need to recalculate the block hash and compare it against the target.